### PR TITLE
Tokenizers Support

### PR DESCRIPTION
**Pull request by tarekgh**
This PR introduces the first version of Tokenizers support. This version will include the following:

- Tokenizer APIs (creation and encoding).
- Abstraction for tokenizers Normalization, pre-processing, and models.
- Bpe model including training support.
-  EnglishRoberta model which is used with the text classification machine learning model.
- Lower and upper casing normalization.
- White space and English roberta pre-processors.
- The integration of the tokenizer with the text classification model.

Features to add later to teh tokenizers:
- More tokenization models (e.g. WordPiece, Unigram,...etc.)
- Save and Load the whole tokenizer. We support saving/loading models but not the whole tokenizer.
- Post-Processing. We support pre-processing only for now. Supporting post processing is good to support too.
- Batch Processing. Support encoding multiple sentences.
- Adding more normalizers and pre-processors.

### CODE REVIEW

**Review by richlander — APPROVED** [review]

File: src/Microsoft.ML.Tokenizers/AddedToken.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        /// Create a new AddedToken object from the given content, specifying if it is intended to be a
+        /// special token. Special tokens are not normalized by default.
+        /// </summary>
+        /// <param name="content">The content of the added token.</param>
+        /// <param name="special">Indicate whether this token is special.</param>
+        public AddedToken(string content, bool special = false) : this()
+        {
+            Content = content ?? "";
+            Special = special;
+            Normalized = !special;
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Tokenizers/AddedToken.cs:70]: Will we ever want a normalized and special token? I'm not really sure what the use case of a special token is so I'm just curious.

File: src/Microsoft.ML.Tokenizers/Microsoft.ML.Tokenizers.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  <Import Project="$(RepoRoot)eng/pkg/Pack.props" />
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <PackageDescription>Microsoft.ML.Tokenizers contains the implmentation of the tokenization used in the NLP transforms.</PackageDescription>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.Text.Json" Version="6.0.5" />
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Tokenizers/Microsoft.ML.Tokenizers.csproj:11]: Can we use a variable from the `versions.props` file so we can keep all the versions in a single place?

File: src/Microsoft.ML.Tokenizers/Model/BPE.cs [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+            return WordToTokens(ref word);
+        }
+
+        /// <summary>
+        /// Map the token to tokenized Id.
+        /// </summary>
+        /// <param name="token">The token to map to the Id.</param>
+        /// <returns>The mapped Id of the token.</returns>
+        public override int? TokenToId(string token)
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Tokenizers/Model/BPE.cs:172]: Is there a reason this takes a string when we have an actual `Token` object?

Comment (michaelgsharp) [reply src/Microsoft.ML.Tokenizers/Model/BPE.cs:172]: Nevermind, I see the implicit conversion and that `Vocab` is from string to int.

File: src/Microsoft.ML.Tokenizers/PreTokenizer/Roberta.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        /// Gets a singleton instance of the Roberta pre-tokenizer..
+        /// </summary>
+        public static readonly RobertaPreTokenizer Instance = new RobertaPreTokenizer();
+
+        private const string Pattern = @"'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+";
+
+        private static readonly IReadOnlyList<Split> _emptyList = new List<Split>();
+
+        /// <summary>
+        /// Splits the given string in multiple substrings at teh word boundary, keeping track of the offsets of said substrings from the original string.
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Tokenizers/PreTokenizer/Roberta.cs:27]: nit: `at the word boundary`

Comment (michaelgsharp) [reply src/Microsoft.ML.Tokenizers/PreTokenizer/Roberta.cs:27]: This is also in the other "PreTokenizer" classes.

File: src/Microsoft.ML.Tokenizers/Utils/Vec.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Tokenizers
+{
+    internal struct Vec<T>
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Tokenizers/Utils/Vec.cs:11]: Is the reason you aren't using a list to get this to be a struct?

**Review by michaelgsharp — APPROVED** [review]
Couple of minor questions but otherwise it looks good to me.

**Comment by michaelgsharp** [comment]
We are also going to release the tokenizers as their own nuget package right?

File: src/Microsoft.ML.Tokenizers/AddedToken.cs (hunk 1 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/AddedToken.cs:70]: It doesn't have to be a special token. You can add any token. The [link](https://huggingface.co/docs/tokenizers/api/added-tokens#tokenizers.AddedToken.normalized) has more info and example why normalization can be needed for added tokens.

File: src/Microsoft.ML.Tokenizers/Microsoft.ML.Tokenizers.csproj (hunk 2 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Microsoft.ML.Tokenizers.csproj:11]: Good idea. Let me try that.

File: src/Microsoft.ML.Tokenizers/PreTokenizer/Roberta.cs (hunk 4 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/PreTokenizer/Roberta.cs:27]: good catch :-)

File: src/Microsoft.ML.Tokenizers/Utils/Vec.cs (hunk 5 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Utils/Vec.cs:11]: Vec exists for a perf reason in some scenarios. 
- Vec doesn't allocate any memory by default when start using it. As it is a struct not allocating any object and not even allocating any underlying buffer till it gets used. 
- The more important reason is it allows modifying the item inside the list without the need for copying. Here is an example:

```C#
            List<MyStruct> list = new();
            list.Add(new MyStruct() { X = 10, Y = 20 });
            list[0].X = 30; // <<<<<<<<<< You cannot do that with the list, but you can do it with the Vec.
```

**Comment by tarekgh** [comment]
> We are also going to release the tokenizers as their own nuget package right?

Yes, I have included the following item inside the [csproj](https://github.com/dotnet/machinelearning/pull/6272/files#diff-9a28487ad7f2cc3b7d3d222130f7b53881424ef0a7651b1262fff5d9ba669b5d):

```Xml
<Import Project="$(RepoRoot)eng/pkg/Pack.props" />
```

Isn't good enough to produce its own package? or do I need to do anything more?

**Comment by tarekgh** [comment]
@michaelgsharp this failure is unrelated, I am wondering if it is a known issue we need to fix.

https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-machinelearning-refs-pull-6272-merge-e0b23e38b82d4f0199/Microsoft.ML.AutoML.Tests/1/console.610df26f.log?helixlogtype=result
```
Starting test: Microsoft.ML.AutoML.Test.AutoFitTests.AutoFitContextLogTest
    Microsoft.ML.AutoML.Test.AutoFitTests.AutoFitRankingCVTest [FAIL]
      Assert.True() Failure
      Expected: True
      Actual:   False
      Stack Trace:
        D:\a\1\s\test\Microsoft.ML.AutoML.Tests\AutoFitTests.cs(351,0): at Microsoft.ML.AutoML.Test.AutoFitTests.AutoFitRankingCVTest()
```

**Comment by michaelgsharp** [comment]
If I remember correctly yes this is one of the ones that is flaky every once and a while. I'll requeue, but we should be good to merge.

**Pull request merged**